### PR TITLE
End to end test example

### PR DESCRIPTION
**Pull request by slinkydeveloper**
Fix https://github.com/restatedev/node-template-generator/issues/18 and https://github.com/restatedev/sdk-typescript/issues/101, replaces https://github.com/restatedev/node-template-generator/pull/19

### CODE REVIEW

File: typescript/end-to-end-testing/test/test.ts [hunk 1]
```diff
@@ -0,0 +1,142 @@
...
+import { createConnectTransport } from "@connectrpc/connect-node";
+import { ExampleService } from "./generated/example_connect";
+import * as http2 from "http2";
+import * as net from "net";
+
+import { protoMetadata } from "../src/generated/proto/example";
+import { MyExampleService } from "../src/example_service";
+
+// Prepare the restate server
+async function prepareRestateServer(
```
Comment (gvdongen) [inline typescript/end-to-end-testing/test/test.ts:19]: Can we pull these out into another file so users could basically copy over the file and write their own tests?

File: typescript/end-to-end-testing/test/test.ts [hunk 2]
```diff
@@ -0,0 +1,142 @@
...
+  }
+}
+
+describe("ExampleService", () => {
+  let startedRestateHttpServer: http2.Http2Server;
+  let startedRestateContainer: StartedTestContainer;
+  let clientTransport: Transport;
+
+  // Deploy Restate and the Service endpoint once for all the tests in this suite
+  beforeAll(async () => {
```
Comment (gvdongen) [inline typescript/end-to-end-testing/test/test.ts:106]: Can we abstract this away so users can just copy over some test_setup.ts file, containing all this logic,
and then in their test just do something like: `beforeAll (async () => {clientTransport = bringUpSetup()})` ?

And `afterAll(async () => {await bringDownSetup()})` ?
This way we could easily document it as well.

**Review by gvdongen — COMMENTED** [review]
Looks good, will be a nice addition! I think we can make it slightly easier to be replicated by having two files for the testing logic: one to copy over with the logic of setting everything up (later to be put into an npm package), and one which just shows a single test that then uses the setup. What do you think?
Also, can we add an example for the handler API as well?

File: typescript/end-to-end-testing/test/test.ts (hunk 1 above)
Comment (slinkydeveloper) [reply typescript/end-to-end-testing/test/test.ts:19]: Can do

**Comment by slinkydeveloper** [comment]
> Also, can we add an example for the handler API as well?

Can I do this in a followup?

File: README.md [hunk 3]
```diff
@@ -68,6 +68,11 @@ wget https://github.com/restatedev/examples/releases/latest/download/kotlin-hell
 wget https://github.com/restatedev/examples/releases/latest/download/typescript-payment-api.zip && unzip typescript-payment-api.zip -d typescript-payment-api && rm typescript-payment-api.zip
 ```
 
+[End to End testing](typescript/end-to-end-testing): Example of how to test End to End Restate services
```
Comment (gvdongen) [inline README.md:71]:
```suggestion
[End-to-end testing](typescript/end-to-end-testing): Example of how to test Restate services end-to-end
```

File: typescript/end-to-end-testing/README.md [hunk 4]
```diff
@@ -0,0 +1,25 @@
...
+
+## Download the example
+
+```shell
+wget https://github.com/restatedev/examples/releases/latest/download/typescript-end-to-end-testing.zip && unzip typescript-end-to-end-testing.zip -d typescript-end-to-end-testing && rm typescript-end-to-end-testing.zip
+```
+
+## Run
+
+Run the `buf` code generation first:
```
Comment (gvdongen) [inline typescript/end-to-end-testing/README.md:15]:
```suggestion
Install the dependencies and run the `buf` code generation first:
```

File: typescript/end-to-end-testing/README.md [hunk 5]
```diff
@@ -0,0 +1,25 @@
...
+
+```shell
+npm run proto
```
Comment (gvdongen) [inline typescript/end-to-end-testing/README.md:18]:
```suggestion
npm install & npm run proto
```

**Review by gvdongen — COMMENTED** [review]
Thanks for splitting up! Much nicer now. 
I get the following error when running `npm run test`
```
 FAIL  test/test.ts
  ● ExampleService › works

    Error 404 during registration:

      79 |     if (!res.ok) {
      80 |       const badResponse = await res.text();
    > 81 |       throw new Error(
         |             ^
      82 |         `Error ${res.status} during registration: ${badResponse}`
      83 |       );
      84 |     }

      at prepareRestateTestContainer (test/restate_test_environment.ts:81:13)
      at Function.start (test/restate_test_environment.ts:121:35)
      at Object.<anonymous> (test/test.ts:15:30)


  ● Test suite failed to run

    TypeError: Cannot read properties of undefined (reading 'stop')

      31 |   // Stop Restate and the Service endpoint
      32 |   afterAll(async () => {
    > 33 |     await restateTestEnvironment.stop();
         |                                  ^
      34 |   });
      35 |
      36 |   it("works", async () => {

      at Object.<anonymous> (test/test.ts:33:34)

Test Suites: 1 failed, 1 total
Tests:       1 failed, 1 total
Snapshots:   0 total
Time:        3.987 s, estimated 6 s
Ran all test suites.

Jest has detected the following 1 open handle potentially keeping Jest from exiting:

  ●  TCPSERVERWRAP

      22 |   await new Promise((resolve, reject) => {
      23 |     restateHttpServer
    > 24 |       .listen(0)
         |        ^
      25 |       .once("listening", resolve)
      26 |       .once("error", reject);
      27 |   });

      at test/restate_test_environment.ts:24:8
      at prepareRestateServer (test/restate_test_environment.ts:22:9)
      at Function.start (test/restate_test_environment.ts:120:42)
      at Object.<anonymous> (test/test.ts:15:59)
```

**Comment by slinkydeveloper** [comment]
>   Error 404 during registration:
    
Perhaps you have an old image?

**Comment by gvdongen** [comment]
@slinkydeveloper That was indeed the issue. 
Should we just for sake of documentation add this to the readme? Something like:
```
If you are encountering errors, try to do `docker pull docker.io/restatedev/restate:latest`.
```
I guess if I bump into this, I am not the only one.

**Comment by gvdongen** [comment]
LGTM!

**Pull request merged**